util/dvsim/utils.py

```diff
@@ -86,84 +95,280 @@ def parse_hjson(hjson_file):
+                any_err = True
+                continue
+
+            raise ValueError('String contains circular expansion of '
+                             'wildcard {!r}.'
+                             .format(match.group(0)))
+
+        # Treat eval_cmd specially
+        if name == 'eval_cmd':
+            cmd = _subst_wildcards(right_str[match.end():],
+                                   mdict, ignored, ignore_error, seen)[0]
+
+            # Are there any wildcards left in cmd? If not, we can run the
+            # command and we're done.
+            cmd_matches = list(wildcard_re.finditer(cmd))
+            if not cmd_matches:
+                var = var[:idx] + run_cmd(cmd)
+                continue
+
+            # Otherwise, check that each of them is ignored, or that
+            # ignore_error is True.
+            bad_names = False
+            if not ignore_error:
+                for cmd_match in cmd_matches:
+                    if cmd_match.group(1) not in ignored:
+                        bad_names = True
+
+            if bad_names:
+                raise ValueError('Cannot run eval_cmd because the command '
+                                 'expands to {!r}, which still contains a '
+                                 'wildcard.'
+                                 .format(cmd))
+
+            # We can't run the command (because it still has wildcards), but we
+            # don't want to report an error either because ignore_error is true
+            # or because each wildcard that's left is ignored. Return the
+            # partially evaluated version.
+            return (var[:idx] + right_str[:match.end()] + cmd, True)
+
+        # Otherwise, look up name in mdict.
+        value = mdict.get(name)
+
+        # If the value isn't set, check the environment
+        if value is None:
+            value = os.environ.get(name)
+
+        if value is None:
+            # Ignore missing values if ignore_error is True.
+            if ignore_error:
+                idx += match.end()
+                continue
+
+            raise ValueError('String to be expanded contains '
+                             'unknown wildcard, {!r}.'
+                             .format(match.group(0)))
+
+        value = _stringify_wildcard_value(value)
+
+        # Do any recursive expansion of value, adding name to seen (to avoid
+        # circular recursion).
+        value, saw_err = _subst_wildcards(value, mdict,
+                                          ignored, ignore_error, seen + [name])
+
+        # Replace the original match with the result and go around again. If
+        # saw_err, increment idx past what we just inserted.
+        var = (var[:idx] +
+               right_str[:match.start()] + value + right_str[match.end():])
+        if saw_err:
+            any_err = True
+            idx += match.start() + len(value)
+
+
+def subst_wildcards(var, mdict, ignored_wildcards=[], ignore_error=False):
```

Comment:
Externally, the key for which the substitutions on `var` is sought (at the `subst_wildcards` call site) should be preemptively added to the `seen` list.

Reply:
That's a good point. As the code currently stands, if you pass something like `{'a': '{a}'}` to `find_and_substitute_wildcards` then `_subst_wildcards` runs twice with `var` equal to `{a}`, but with `a` in the seen list on the second time (at which point it explodes). I hadn't noticed this! But I'm not convinced it really matters: we still spot the circular reference and just take one more iteration to do so.

The problem with your proposed change is that this leaks the existence of `seen` (an implementation detail) in the API. Since it's not actually needed for correctness, I think I'd prefer to keep the simpler API. What do you think?

Reply:
Sure, that's fine. We can deal with it later.